BUAN bundle highlight option in horizon

BramshQamar opened the pull request:
Added functionality in `dipy_horizon` to show highlighted areas/segments on a bundle where pvalues (negative log of pvalues) are greater than a certain threshold (threshold is a parameter).

Usage:

`dipy_horizon IFOF_L.trk IFOF_L_fa_pvalues_log.npy --buan --buan_thr 2`

`buan_thr 2` for `-log(pvalues)` is equal to `0.01` pvalues.


Resulting visualization looks something like this:

<img width="597" alt="Screen Shot 2020-06-10 at 10 34 19 PM" src="https://user-images.githubusercontent.com/26060471/84338791-01e72f80-ab6b-11ea-9a9d-cd6d32d40e9a.png">


Where, red color indicates -log(pvalues) greater than a buan threshold of 2.

We can also provide `pvalues` instead of `-log(pvalues)`.

`dipy_horizon IFOF_L.trk IFOF_L_fa_pvalues.npy --buan --buan_thr 0.01`

Results will be the same but with inverted colors. 
<img width="864" alt="Screen Shot 2020-06-10 at 10 33 47 PM" src="https://user-images.githubusercontent.com/26060471/84338705-c2b8de80-ab6a-11ea-9cb5-34c8cbd987f7.png">
Now, green is the highlighted area.

pep8speaks commented:
Hello @BramshQamar, Thank you for submitting the Pull Request !



* In the file [`dipy/workflows/viz.py`](https://github.com/dipy/dipy/blob/efb125c81f5039bdcb9fd9eeee6eececeb4e0353/dipy/workflows/viz.py):

> [Line 184:41](https://github.com/dipy/dipy/blob/efb125c81f5039bdcb9fd9eeee6eececeb4e0353/dipy/workflows/viz.py#L184): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 184:43](https://github.com/dipy/dipy/blob/efb125c81f5039bdcb9fd9eeee6eececeb4e0353/dipy/workflows/viz.py#L184): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 186:41](https://github.com/dipy/dipy/blob/efb125c81f5039bdcb9fd9eeee6eececeb4e0353/dipy/workflows/viz.py#L186): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 186:43](https://github.com/dipy/dipy/blob/efb125c81f5039bdcb9fd9eeee6eececeb4e0353/dipy/workflows/viz.py#L186): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

Do see the [DIPY coding Style guideline](https://github.com/dipy/dipy/blob/master/doc/devel/coding_style_guideline.rst)

BramshQamar commented:
Hello @Garyfallidis and @skoudoro,

This PR is ready and needs review.

The only test failing on all machines is `FAILED workflows/tests/test_reconst_mapmri.py::test_reconst_mmri_both `. Which is not related to this PR.

Garyfallidis commented on dipy/viz/app.py:
```diff
@@ -775,6 +792,9 @@ def horizon(tractograms=None, images=None, pams=None,
         Each tuple contains data and affine
     pams : sequence of PeakAndMetrics
         Contains peak directions and spherical harmonic coefficients
+    np_files : sequence of arrays
```
``np_files`` seems too generic to me if it is going to be used only for BUAN.

Garyfallidis replied:
But I think you kept that name because it can be used for other visualizations too.
Maybe ``npy_files``?

Garyfallidis commented on dipy/workflows/viz.py:
```diff
@@ -150,6 +158,36 @@ def run(self, input_files, cluster=False, cluster_thr=15.,
+            if ends(".npy"):
+
+                data = np.load(fname)
+                numpy_files.append(data)
```
I am not sure how safe is appending numpy arrays into lists. I think you should use ``ArraySequence`` if you can. That is much preferred method. Remember that we had memory issues when adding arrays to lists. You can check if this problem persists by trying to delete this memory and see if the memory (list of arrays) is actually being removed from the RAM while the object is running.

BramshQamar replied:
These files are very small in size. Eg: 100 integers only in one file. Mostly user will send one .npy file with pvalues in it (100 integers in the file). This should not cause any memory issues.

Garyfallidis commented on dipy/workflows/viz.py:
```diff
@@ -150,6 +158,36 @@ def run(self, input_files, cluster=False, cluster_thr=15.,
+                colors = []
+                for i in range(len(ind)):
+
+                    if pvalues[ind[i]] < buan_thr:
```
I would prefer if the highlight color was available as an option.
Also I would like if we could keep the original color of the bundle if we want.

BramshQamar replied:
Done. Added original color for bundles with bundle highlight area color as an option (default red).

skoudoro commented:
Hi @BramshQamar,

Please, can you rebase your PR? Also, can you address @Garyfallidis comments and add some tests. Thank you

BramshQamar commented:
Updates this PR color feature. 

- [ ] - Bundles now have an original color with bundle highlight area color as optional input parameter (default red).

- [ ] - Giving both p-values and log(p-values) was redundant. It only takes p-values now. 

Some examples:
<img width="828" alt="Screen Shot 2020-10-15 at 4 08 17 PM" src="https://user-images.githubusercontent.com/26060471/96181152-8bd04c00-0f01-11eb-9cde-29df6e3a472f.png">

<img width="449" alt="Screen Shot 2020-10-15 at 4 10 39 PM" src="https://user-images.githubusercontent.com/26060471/96181172-968ae100-0f01-11eb-99db-4af0a82f8610.png">

- [ ] Needs tests

BramshQamar commented:
> Hi @BramshQamar,
> 
> Please, can you rebase your PR? Also, can you address @Garyfallidis comments and add some tests. Thank you

Hi @skoudoro, I have addressed all @Garyfallidis comments and added tests as well.
Let me know if I missed anything. Thank you!

skoudoro commented on dipy/viz/app.py:
```diff
@@ -282,15 +291,21 @@ def add_cluster_actors(self, scene, tractograms,
                     apply_shader(self, centroid_actor)
 
             else:
+                if self.buan:
+                    streamline_actor = actor.line(streamlines,
+                                                  self.buan_colors[color_ind])
+                else:
+                    streamline_actor = actor.line(streamlines, colors=colors)
```
Just a small suggestion: you could replace this block with this 2 lines:

```python
s_color = self.buan_colors[color_ind] if self.buan else colors
streamline_actor = actor.line(streamlines, colors=s_colors)
```

skoudoro commented on dipy/viz/app.py:
```diff
@@ -807,6 +823,10 @@ def horizon(tractograms=None, images=None, pams=None,
     interactive : bool
         Allow user interaction. If False then Horizon goes on stealth mode
         and just saves pictures.
+    buan : bool
```
`buan : bool, optional` instead

skoudoro commented on dipy/viz/app.py:
```diff
@@ -807,6 +823,10 @@ def horizon(tractograms=None, images=None, pams=None,
         Allow user interaction. If False then Horizon goes on stealth mode
         and just saves pictures.
+    buan : bool
+        Enables BUAN framework visualization.
```
`Default is False`, at the end

skoudoro commented on dipy/workflows/tests/test_viz.py:
```diff
@@ -67,6 +67,13 @@ def test_horizon_flow():
 
+    buan_colors = np.ones(streamlines.get_data().shape)
+
+    print(buan_colors)
```
`print` to remove

skoudoro commented on dipy/workflows/viz.py:
```diff
@@ -6,7 +6,15 @@
+
+if has_fury:
+    from fury.colormap import line_colors
+from fury.utils import numpy_to_vtk_colors
```
indentation issue

skoudoro commented on dipy/workflows/viz.py:
```diff
@@ -62,6 +70,15 @@ def run(self, input_files, cluster=False, cluster_thr=15.,
         disable_order_transparency : bool, optional
             Default False. Use depth peeling to sort transparent objects.
             If True also enables anti-aliasing.
+        buan : bool
```
`buan : bool, optional` instead

skoudoro commented on dipy/workflows/viz.py:
```diff
@@ -62,6 +70,15 @@ def run(self, input_files, cluster=False, cluster_thr=15.,
             Default False. Use depth peeling to sort transparent objects.
             If True also enables anti-aliasing.
+        buan : bool
+            Enables BUAN framework visualization.
```
same as before `default...`

skoudoro reviewed (commented):
Hi @BramshQamar,

Thank you for the update!  Overall, it looks good. There are just some quick details to fix. I will try your code a bit later in the day

skoudoro commented:
Also, Can you fix all pep8 listed [here](https://github.com/dipy/dipy/pull/2181#issuecomment-642366335)? 
Thank you

skoudoro commented:
Thank you @BramshQamar! merging
